Technology specific discount rates

robertodawid opened the pull request:
Changes in code to solve issue #61

Hello, I have made a few changes to the code. The capital recovery factors use a global discount rate, and the new one _CapitalRecoveryFactorIdv[r, t],_ uses individual discount rates. This factor multiplied by the _Capitalinvest[r,t,y]_ calculates :
•	The annuity payment for n years (assuming that the economic and lifetime of a technology are the same). 
•	then, the PV of those annuities in the investment year at the specific discount rate, 
•	Finally, it is discounted at the global discount rate.

The parameter  _TypeOfDiscountRate{r in REGION}_ allows selecting between equations CC2 and CC3. CC2 calculates the discounted capital investment using ONLY a global discount rate, and CC3 using individual discount rates.

willu47 commented:
Hi @robertodawid - many thanks for this.  @abhishek0208 do have time to take a look at @robertodawid's solution to #61 as I know this was a concern for Tom?

In the meantime, I have a few suggestions which I'll make in a review.

willu47 commented on src/osemosys.txt:
```diff
@@ -75,6 +82,8 @@ param Conversionlh{l in TIMESLICE, lh in DAILYTIMEBRACKET} binary;
 param DaysInDayType{ls in SEASON, ld in DAYTYPE, y in YEAR};
 param TradeRoute {r in REGION, rr in REGION, f in FUEL, y in YEAR} binary;
 param DepreciationMethod{r in REGION};
+param TypeOfDiscountRate{r in REGION};
```
Suggest constraining this to either `0` or `1` using 
```ampl
param TypeOfDiscountRate{r in REGION}, in 0..1;
```

willu47 commented on src/osemosys.txt:
```diff
@@ -562,14 +571,17 @@ s.t. SI10_TotalDiscountedCostByStorage{r in REGION, s in STORAGE, y in YEAR}: Di
+
+s.t. CC2_DiscountingCapitalInvestment{r in REGION, t in TECHNOLOGY, y in YEAR: TypeOfDiscountRate[r]=1}: CapitalInvestment[r,t,y]/(CapitalRecoveryFactor[r,y]) = DiscountedCapitalInvestment[r,t,y];
+
+s.t. CC3_DiscountingCapitalInvestment{r in REGION, t in TECHNOLOGY, y in YEAR: TypeOfDiscountRate[r]=2}: CapitalInvestment[r,t,y]*(CapitalRecoveryFactorIdv[r,t,y]) = DiscountedCapitalInvestment[r,t,y];
```
Do we need this extra equation, or could we just ensure that the `DiscountRateIdv` defaults to the `DiscountRate`? We could then also remove the `TypeOfDiscountRate` parameter.

willu47 commented on src/osemosys.txt:
```diff
@@ -56,11 +56,18 @@ set STORAGE;
+param DiscountRate{r in REGION};
+param DiscountRateIdv{r in REGION, t in TECHNOLOGY};
+
+param CapitalRecoveryFactor{r in REGION, y in YEAR} :=
```
Note issue #60 - really these parameter names are incorrect and should be `DiscountFactor`

willu47 commented on src/osemosys.txt:
```diff
@@ -56,11 +56,18 @@ set STORAGE;
-param CapitalRecoveryFactorMid{r in REGION, t in TECHNOLOGY, y in YEAR} :=
-	(1 + DiscountRate[r, t]) ^ (y - min{yy in YEAR} min(yy) + 0.5);
+param DiscountRate{r in REGION};
+param DiscountRateIdv{r in REGION, t in TECHNOLOGY};
```
Note below suggests setting default to that of `DiscountRate` allowing you to potentially remove CC3 and `TypeOfDiscountRate`

willu47 reviewed (changes requested):
Overall this looks like a good improvement. Thanks a lot! 

However, there are a few further steps before we include this in the next release.

1. A *very* simple example model (see the tests folder) which demonstrates the discount rate working as expected.
2. Implementing for short and fast codes
3. Getting the tests to pass

abhishek0208 commented:
> Hi @robertodawid - many thanks for this. @abhishek0208 do have time to take a look at @robertodawid's solution to #61 as I know this was a concern for Tom?
> 
> In the meantime, I have a few suggestions which I'll make in a review.

@willu47 This was on my to-do list so glad that it's already being done. Glad to give it a review (and run it by Tom too).

abhishek0208 commented:
Hi @robertodawid, I tested out the script with a simple model and it works as expected (i.e. technologies with a lower technology-specific discount rate are preferred to those with a higher rate). However, couldn't the equation be simplified as below:

abhishek0208 commented on src/osemosys.txt:
```diff
@@ -56,11 +56,18 @@ set STORAGE;
+param OperationalLife{r in REGION, t in TECHNOLOGY};
+
+param CapitalRecoveryFactorIdv{r in REGION, t in TECHNOLOGY, y in YEAR} :=  
+	(DiscountRateIdv[r,t] / (1 - (1 + DiscountRateIdv[r,t])^(-(OperationalLife[r,t])))) * ((1 - (1 + DiscountRateIdv[r,t])^(-(OperationalLife[r,t]))) * (1 + DiscountRateIdv[r,t]) / DiscountRateIdv[r,t]) / ((1 + DiscountRate[r]) ^ (y - min{yy in YEAR} min(yy) + 0.0));
```
Couldn't this equation be simplified as below?
`(1 + DiscountRateIdv[r,t]) / ((1 + DiscountRate[r]) ^ (y - min{yy in YEAR} min(yy) + 0.0))`

robertodawid replied:
Hi, Abhi, thanks for testing the script. Yes, it is possible to simplify that eq.. with (1+DR). I left it as a long eq. , so others can see where the different factors come from.

willu47 commented:
Thanks @abhishek0208 for performing the review.  @robertodawid - please could you make the changes suggested by @abhishek0208 and me and push the commits to the same branch. This will update the pull request. Once the tests pass (which will require the same change made to short and fast versions of the code), we'll be able to merge in your changes and make a new release!

robertodawid commented:
Hi, @willu47, I have updated the pull request. Thanks, @abhishek0208 for your help. These are the main updates:

1.	From #60. Changed parameter names ```CapitalRecoveryFactor``` for ```DiscountFactor``` 
2.	Deleted parameter ```TypeofDiscountRate``` and CC3 removed
3.	Updated CC1 and CC2 
4.	Parameter ```DiscountRateIdv``` defaults to ```DiscountRate``` 
```param DiscountRateIdv{r in REGION, t in TECHNOLOGY}, default DiscountRate[r];```

5.	Added ```CapitalRecoveryFactor``` (CRF).
This factor allows calculating the individual payments from a lump sum (investment cost) every year of a power plant's operational life. CRF is calculated by using the individual discount rate. The individual discount rates are used only in the CRF.

```param CapitalRecoveryFactor{r in REGION, t in TECHNOLOGY} := (1 - (1 + DiscountRateIdv[r,t])^(-1))/(1 - (1 + DiscountRateIdv[r,t])^(-(OperationalLife[r,t])));```


6.	Added ```PvAnnuity```
This factor allows calculating the present value of the all-individual payments (calculated by the CRF) in the investment year at the global discount rate.

```param PvAnnuity{r in REGION, t in TECHNOLOGY} := (1 - (1 + DiscountRate[r])^(-(OperationalLife[r,t]))) * (1 + DiscountRate[r]) / DiscountRate[r];```

I have updated the long, short, and fast versions of the code and tested them with the Utopia data file.

willu47 commented:
@abhishek0208 - any chance you could finish your review of the revised code?

abhishek0208 commented:
@willu47 @robertodawid sorry for the delay. Tested out all three versions of `osemosys` with `utopia.txt` and it looks good. Ready to be merged now. I suppose `otoole` will need to be updated? I can try and implement that if it helps.
